pkg/apis/certmanager/v1/types_issuer.go

```diff
@@ -269,13 +269,34 @@ type VaultKubernetesAuth struct {
 	// The required Secret field containing a Kubernetes ServiceAccount JWT used
 	// for authenticating with Vault. Use of 'ambient credentials' is not
 	// supported.
-	SecretRef cmmeta.SecretKeySelector `json:"secretRef"`
+	// +optional
+	SecretRef cmmeta.SecretKeySelector `json:"secretRef,omitempty"`
+	// Note: it should be a pointer because it is optional. However, for
+	// backward compatibility, we cannot change it to a pointer.
```

Comment:
**question to the reviewer:** I am unsure about what to do with `SecretRef`.

Since this field becomes optional, the [api-conventions.md](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required) states that I should be using a pointer.

On the other side, changing from

```go
SecretRef cmmeta.SecretKeySelector `json:"secretRef,omitempty"`
```

to

```go
SecretRef *cmmeta.SecretKeySelector `json:"secretRef,omitempty"`
```

means that I will be breaking anyone importing this package.

I think it is fine to not use a pointer to keep backwards compatibility. I added a note to keep that in mind, since I think the inconsistency will confuse people, i.e., this:


```go
// +optional
SecretRef cmmeta.SecretKeySelector `json:"secretRef,omitempty"`

// +optional
ServiceAccountRef *ServiceAccountRef `json:"serviceAccountRef,omitempty"`
```

Reply:
I think that changing the field to a pointer does not break backwards compatibility towards people creating issuers from YAML manifests.

Reply:
You are correct, the YAML API won't be affected.

That said, the Go API will change, and anyone relying on this `SecretRef` struct field will break.

That said:

1. The cert-manager API is not often used as a transitive dependency (if it was, it would cause much more trouble as in https://maelvls.dev/client-go-transitive-hell/).
2. The field `SecretRef` isn't usually used when importing the cert-manager API. Usually, people use the certificaterequest API for implementing an issuer.

I did a GitHub search and can't find anyone using `Issuer.spec.vault.kubernetes.secretRef`.

For these two reasons, I propose to make that breaking change anyways.

Reply:
We do specifically call out `pkg/apis/` as being something that people should be safe to import in our [relevant docs](https://cert-manager.io/docs/contributing/importing/#when-you-might-import-cert-manager) on the topic.

I'm inclined to think we can't ever break anything under `pkg/apis/` without bumping the import path to cert-manager/v2 to be honest.

Reply:
Agreed. I think it's fine to add a comment below SecretRef to explain why it isn't a pointer, and then to adjust the validation logic. 

And if we move to CEL one day, it will still be possible to validate that field. It will be a bit trickier, but still possible.

Reply:
I changed it back to a non-pointer type.